File: j-LIB-core/src/main/java/io/github/dot166/jLib/app/Notifier.java
package io.github.dot166.jLib.app;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;

public abstract class Notifier {
    private final NotificationManager notificationManager;

    public Notifier(NotificationManager notificationManager) {
        this.notificationManager = notificationManager;
    }

    public abstract String getNotificationChannelId();
    public abstract String getNotificationChannelName();
    public abstract int getNotificationId();

    public void showNotification() {
        NotificationChannel channel = createNotificationChannel();
        notificationManager.createNotificationChannel(channel);
        Notification notification = buildNotification();
        notificationManager.notify(
                getNotificationId(),
                notification
        );
    }

    public NotificationChannel createNotificationChannel(int importance) {
        return new NotificationChannel(
                getNotificationChannelId(),
                getNotificationChannelName(),
                importance
        );
    }

    public NotificationChannel createNotificationChannel() {
        return createNotificationChannel(NotificationManager.IMPORTANCE_DEFAULT);
    }

    public abstract Notification buildNotification();

    protected abstract String getNotificationTitle();

    protected abstract String getNotificationMessage();
}

File: j-LIB-core/src/main/java/io/github/dot166/jlib/app/Notifier.kt
package io.github.dot166.jlib.app

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager

abstract class Notifier(private val notificationManager: NotificationManager) {
    abstract val notificationChannelId: String?
    abstract val notificationChannelName: String?
    abstract val notificationId: Int

    open fun showNotification() {
        val channel = createNotificationChannel()
        notificationManager.createNotificationChannel(channel)
        val notification = buildNotification()
        notificationManager.notify(
            this.notificationId,
            notification
        )
    }

    @JvmOverloads
    fun createNotificationChannel(importance: Int = NotificationManager.IMPORTANCE_DEFAULT): NotificationChannel {
        return NotificationChannel(
            this.notificationChannelId,
            this.notificationChannelName,
            importance
        )
    }

    abstract fun buildNotification(): Notification?

    protected abstract val notificationTitle: String?

    protected abstract val notificationMessage: String?
}


